dace/libraries/blas/nodes/gemm.py

```diff
@@ -425,6 +425,575 @@ def expansion(node, state, sdfg):
+            # (which is the latency of a multiply-add + latency for reading and writing to BRAM).
+
+            entry_pipeline, exit_pipeline = state.add_pipeline(
+                "compute_and_drain", {
+                    "n0": f"0:ceiling({N}/{P})",
+                    "tm": f"0:ceiling({M}/{T})",
+                    "k": f"0:{K}",
+                    "m": f"0:{T} + {L}"
+                },
+                drain_size=P * T,
+                drain_overlap=False,
+                additional_iterators={
+                    'm_drain': 0,
+                    'k_drain': 0
+                },
+                schedule=dace.ScheduleType.FPGA_Device)
+
+            # Instantiate buffers
+            sdfg.add_scalar("A_reg",
+                            dtype=dtype_a,
+                            transient=True,
+                            storage=dace.dtypes.StorageType.FPGA_Registers)
+            A_reg = state.add_write("A_reg")
+            A_reg_init = state.add_access("A_reg")
+
+            # For C result we are going to use vectorized data type
+
+            # Note: for some of the Sacred Mysteries of Intel OpenCL Compiler (TM), if this buffer is smaller
+            # than 24 floats, the II of the pipeline will be 5. Therefore we check this and in case we enlarge it
+            buffer_size = T if T_constant is None else max(T_constant, 24)
+            sdfg.add_array("C_buffer", [buffer_size],
+                           dtype=vec_type,
+                           transient=True,
+                           storage=dace.dtypes.StorageType.FPGA_Local)
+            C_buffer_in = state.add_read("C_buffer")
+            C_buffer_out = state.add_write("C_buffer")
+
+            # Init data to reset partial results
+            new_sdfg.add_array("C_init",
+                               dtype=vec_type,
+                               shape=[1],
+                               transient=True,
+                               storage=dace.dtypes.StorageType.FPGA_Registers)
+            C_init = state.add_access("C_init")
+            C_init_tasklet = state.add_tasklet("C_data_init", {}, {"init_data"},
+                                               "init_data = 0")
+
+            state.add_memlet_path(C_init_tasklet,
+                                  C_init,
+                                  src_conn="init_data",
+                                  memlet=dace.Memlet("C_init[0]"))
+            state.add_memlet_path(entry_pipeline,
+                                  C_init_tasklet,
+                                  memlet=dace.Memlet())
+
+            # Feed A
+            # every PE: reads input data, buffer the data assigned to it
+            buffer_a_tasklet = state.add_tasklet(
+                "buffer_a", {"a_in"}, {
+                    "a_reg",
+                }, f"""\
+if m == 0 and not {entry_pipeline.pipeline.drain_condition()}:
+    a_reg = a_in""")
```

Comment:
This is one of the most disgusting `yapf` outputs I have ever seen 😢

Reply:
:(